src/coreclr/jit/gentree.cpp

```diff
@@ -16927,13 +16927,20 @@ void Compiler::gtExtractSideEffList(GenTree*     expr,
 
         fgWalkResult PreOrderVisit(GenTree** use, GenTree* user)
         {
-            GenTree* node = *use;
+            GenTree*     node  = *use;
+            GenTreeFlags flags = m_flags;
+
+            if ((user != nullptr) && (user->OperIs(GT_COMMA)) && (user->AsOp()->gtGetOp1() == node))
+            {
+                // If this is `op1` is a GT_COMMA node, make sure to extract the side-effects of it.
```

Comment:
This needs a better comment on _why_. `COMMA`s are not special. `MUL(x, 0)` is also a node will not include `x` in its VN.

As far as I understand, the problem in the [example](https://github.com/dotnet/runtime/pull/96077#issuecomment-1858644484) is that the two static base helpers are for different classes. The reason cctor helpers are not extracted for CSE purposes is because CSE defs dominate CSE uses, so it is assumed that any cctor side effects will have been executed by the time we get to the use. This reasoning is not fully correct when the use's VN does not capture the full extent of its cctor side effects - you would expect the dropped cctor helper to have something like an exception set describing a different "exception" than the one being CSEd. It doesn't today.

So this looks like a workaround for that bug.

Reply:
> As far as I understand, the problem in the [example](https://github.com/dotnet/runtime/pull/96077#issuecomment-1858644484) is that the two static base helpers are for different classes.

They are not, they are for same class and I include the `CORINFO_HELP_READYTORUN_GCSTATIC_BASE ` during importing of `CORINFO_HELP_READYTORUN_THREADSTATIC_BASE `.

> I am not sure why the `COMMA`'s VN is the same as its op2

Seems like that's how we assign VN for Comma. `op1`'s VN for exception VN and that of `op2` for normal value.

https://github.com/dotnet/runtime/blob/187406e807b0b1d2ae76ed55f181b28abcdf052b/src/coreclr/jit/valuenum.cpp#L11442-L11447

Reply:
> They are not, they are for same class

Hmm, but then why is it a problem that it is removed?

Reply:
I think the issue here is that in #89472's expansion

```
N007 ( 45, 19) [000008] --CXG------                         *  CALL      void   Program:Consume(System.String)
N006 ( 31, 13) [000007] n-CXG------ arg0 in rcx             \--*  IND       ref   
N005 ( 29, 11) [000006] --CXG--N---                            \--*  ADD       byref 
N003 ( 28, 10) [000004] --CXG------                               +--*  COMMA     byref 
N001 ( 14,  5) [000001] H-CXG------                               |  +--*  CALL help byref  CORINFO_HELP_READYTORUN_GCSTATIC_BASE
N002 ( 14,  5) [000000] H-CXG------                               |  \--*  CALL help byref  CORINFO_HELP_READYTORUN_THREADSTATIC_BASE
N004 (  1,  1) [000005] -----------                               \--*  CNS_INT   long   120 Fseq[i]

N001 [000001]   CALL help => $1c0 {norm=$180 {ReadyToRunStaticBaseGC($140)}, exc=$100 {HelperMultipleExc()}}
N002 [000000]   CALL help => $1c1 {norm=$181 {ReadyToRunStaticBaseThread($141)}, exc=$100 {HelperMultipleExc()}}
N003 [000004]   COMMA     => $1c1 {norm=$181 {ReadyToRunStaticBaseThread($141)}, exc=$100 {HelperMultipleExc()}}
N004 [000005]   CNS_INT   120 Fseq[i] => $200 {LngCns:  120}
N005 [000006]   ADD       => $1c3 {norm=$1c2 {ADD($181, $200)}, exc=$100 {HelperMultipleExc()}}
    VNForHandle(i) is $142, fieldType is ref, size = 8
    VNForMapSelect($80, $142):ref returns $101 {$80[$142]}
N006 [000007]   IND       => <l:$103 {norm=$101 {$80[$142]}, exc=$100 {HelperMultipleExc()}}, c:$102 {norm=$240 {MemOpaque:NotInLoop}, exc=$100 {HelperMultipleExc()}}>
  fgCurMemoryVN[GcHeap] assigned for CALL at [000008] to VN: $81.
N007 [000008]   CALL      => $VN.Void
```
The second (`[000000]`) helper call actually can't throw and can't invoke the cctor, so it should not have the same exception set as the first one. So I've suggested to Kunal that he add a NOCTOR variant and use that in the importer.

This would give the COMMA a different VN than the OP2, which would cause CSE to model the COMMA and OP2 independently and avoid the problem where `[000004]`  looks like a use of `[000000]` and leads CSE to try this odd self-extraction.

Reply:
> So I've suggested to Kunal that he add a NOCTOR variant and use that in the importer.

Done that in https://github.com/dotnet/runtime/pull/89472/commits/119e2ae3d890376570fc0a31a2c06381b7b02407 and https://github.com/dotnet/runtime/pull/89472/commits/9cfea3990afafa3204edce8c092518959fa6f9e6 and that seems to solve the problem as seen in generated code I pasted in https://github.com/dotnet/runtime/pull/89472#issuecomment-1862224434. Thanks for your help @AndyAyersMS and @SingleAccretion